dot/sync/unready_blocks.go

```diff
@@ -102,62 +166,46 @@ func (u *unreadyBlocks) isIncomplete(blockHash common.Hash) bool {
 }
 
 // inDisjointFragment iterate through the disjoint fragments and
-// check if the block hash an number already exists in one of them
+// check if the block hash and number already exists in one of them
 func (u *unreadyBlocks) inDisjointFragment(blockHash common.Hash, blockNumber uint) bool {
 	u.mtx.RLock()
 	defer u.mtx.RUnlock()
 
 	for _, frag := range u.disjointFragments {
-		target := &types.BlockData{Header: &types.Header{Number: blockNumber}}
-		idx, found := slices.BinarySearchFunc(frag, target,
-			func(a, b *types.BlockData) int {
-				switch {
-				case a.Header.Number == b.Header.Number:
-					return 0
-				case a.Header.Number < b.Header.Number:
-					return -1
-				default:
-					return 1
-				}
-			})
-
-		if found && frag[idx].Hash == blockHash {
+		bd := frag.Find(func(bd *types.BlockData) bool {
+			return bd.Header.Number == blockNumber && bd.Hash == blockHash
+		})
+
+		if bd != nil {
 			return true
 		}
 	}
 
 	return false
 }
 
-// removeIrrelevantFragments checks if there is blocks in the fragments that can be pruned
-// given the finalised block number
-func (u *unreadyBlocks) removeIrrelevantFragments(finalisedNumber uint) {
+func (u *unreadyBlocks) removeIncompleteBlocks(del func(key common.Hash, value *types.BlockData) bool) {
 	u.mtx.Lock()
 	defer u.mtx.Unlock()
 
-	maps.DeleteFunc(u.incompleteBlocks, func(_ common.Hash, value *types.BlockData) bool {
-		return value.Header.Number <= finalisedNumber
-	})
-
-	fragmentIdx := 0
-	for _, fragment := range u.disjointFragments {
-		// the fragments are sorted in ascending order
-		// starting from the latest item and going backwards
-		// we have a higher chance to find the idx that has
-		// a block with number lower or equal the finalised one
-		idx := len(fragment) - 1
-		for ; idx >= 0; idx-- {
-			if fragment[idx].Header.Number <= finalisedNumber {
-				break
-			}
-		}
+	maps.DeleteFunc(u.incompleteBlocks, del)
+}
 
-		updatedFragment := fragment[idx+1:]
-		if len(updatedFragment) != 0 {
-			u.disjointFragments[fragmentIdx] = updatedFragment
-			fragmentIdx++
-		}
-	}
+// pruneFragments will iterate over the disjoint fragments and check if they
+// can be removed based on the del param
+func (u *unreadyBlocks) pruneDisjointFragments(del func(*Fragment) bool) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
 
-	u.disjointFragments = u.disjointFragments[:fragmentIdx]
+	u.disjointFragments = slices.DeleteFunc(u.disjointFragments, del)
+}
+
+// LowerThanOrEqHighestFinalized returns true if the fragment contains
+// a block that has a number lower than highest finalized number
+func LowerThanOrEqHighestFinalized(highestFinalizedNumber uint) func(*Fragment) bool {
```

Comment:
nit:
```suggestion
// LowerThanOrEq returns a function that returns true if the fragment contains
// a block that has a number lower than or equal to the passed in number
func LowerThanOrEq(blockNumber uint) func(*Fragment) bool {
```
The function is more general than the name and doc comment suggests. It could be called with something other than the highest finalized number.